Bug 1214392: Denormalize latest activity.

Osmose opened the pull request:
This relies on #216, will rebase after that gets merged. Pay attention to the last commit only.

This denormalizes the latest activity such that Project, Resource, Locale, and ProjectLocale (which was previously an implicit through model for Project.locales) all have a `latest_translation` field that gets updated when translations are saved. The listings that show the latest activity now efficiently pull the `latest_translation` field when trying to figure out which translation qualifies as the latest activity, as opposed to doing a bunch of queries to find the latest activity.

Locally this led to a major drop in load time for the project, locale, and resource listings for me. It also reduced the number of database queries on each page by quite a bit.

@mathjazz r?

mathjazz commented:
Thank you for a separate commit! r+

Osmose commented:
Rebased, easier to review now. :P

mathjazz commented on pontoon/base/views.py:
```diff
@@ -138,19 +142,25 @@ def locale_project(request, locale, slug):
+        .filter(resource__project=project, locale=l)
+        .select_related('resource', 'latest_translation')
     )
+    stats = {s.resource.path: s for s in stats_qs}
```
Nice reduction of DB queries!

mathjazz commented on pontoon/base/views.py:
```diff
@@ -138,19 +142,25 @@ def locale_project(request, locale, slug):
-            slug=slug,
-            pk__in=Resource.objects.values('project')
-        )
+    projects = Project.objects.prefetch_related('subpage_set').distinct()
```
Is that distinct() really necessary?

Osmose replied:
Yep, although because I split this into two lines (and I'm not entirely sure if I had a good reason to do so now, but didn't remove it) it's not obvious. The `get_object_or_404` call filters by `resources__isnull`, and anytime you filter by a ManyToMany, you'll get a result for each of the related objects. In this case, we get as many projects as there are resources in the project. Hence, the `distinct` to avoid the duplication.

mathjazz commented on pontoon/base/models.py:
```diff
@@ -152,6 +153,15 @@ class Locale(models.Model):
+        blank=True,
+        null=True,
+        related_name='+',
+        on_delete=models.SET_NULL
```
Woohooo!

mathjazz replied:
That thing up there means I've never seen on_delete before. But I'm also wondering, what would actually change if latest_translation got deleted in a cascade with the foreign key? Seems like you have that case covered in get_latest_activity().

Osmose replied:
It's the other way around: Deleting a translation deletes the locale with the foreign key. This is how I mysteriously deleted projects on my local instance while running a sync, which deleted some translations. 

Suffice it to say that we do _not_ want locales being deleted when translations are deleted. :P

mathjazz commented on pontoon/base/models.py:
```diff
@@ -228,6 +240,15 @@ class Project(models.Model):
+        'Translation',
+        blank=True,
+        null=True,
+        related_name='+',
```
Why do we need no backwards relation? I understand we'll never use it, I'm just wondering why do we want to explicitly prevent it? Optimization?

Osmose replied:
Nah, it's a naming conflict. By default the `related_name` on the Translation for the Locale that it is the `latest_translation` of would be `translation.locale`, but we already have an attribute with that name on the Translation.

And then I copy-pasted it around since, as you said, we'll never use it, but it's only truly necessary on the Locale class.

mathjazz commented:
An error occurs when adding a new locale to a project, but it's easy to fix:
`Cannot set values on a ManyToManyField which specifies an intermediary model.  Use base.ProjectLocale's Manager instead.`

Excellent work, this is a huge improvement, can't wait to deploy this! r+wc

Tested with a few days old production database, following are the average of the 3 page load times before vs after applying the PR:
/projects: 4.5s vs 3.5s
/teams: 4.7s vs 3.5s
/sl: 3.4s vs 1.9s
/projects/pontoon-intro: 4.2s vs 3.3s
/sl/firefox-aurora 6.5s vs 4.1s

Osmose commented:
Rebased and fixed the locale issue. I'm not worried about tests since the admin is going to be killed soon anyway.

We may want to hold off on merging this if we still intend to revert the other two commits this relies on, though.

mathjazz commented:
r+

Not merging until we figure out what to do with reverted commits this relies on.
